ocaml/xapi/xapi_mgmt_iface.ml

```diff
@@ -48,114 +30,151 @@ let update_mh_info interface =
-        match !stunnel_accept with
-        | None ->
-            fun () ->
-              D.warn
-                "reconfigure_stunnel: accept is not set, so not restarting \
-                 stunnel"
-        | Some accept ->
-            fun () -> _restart_stunnel_no_cache ~__context ~accept)
-  in
-  f ()
-
-let stop () =
-  debug "Shutting down the old management interface (if any)" ;
-  List.iter (fun i -> Http_svr.stop i) !management_interface_server ;
-  management_interface_server := [] ;
-  listening_all := false ;
-  listening_localhost := false ;
-  listening_himn := false
-
-(* Even though xapi listens on all IP addresses, there is still an interface appointed as
- * _the_ management interface. Slaves in a pool use the IP address of this interface to connect
- * the pool master. *)
-let start ~__context ?addr () =
-  let socket, accept =
-    match addr with
-    | None -> (
-        info "Starting new server (listening on all IP addresses)" ;
-        try
-          (* Is it IPv6 ? *)
-          let addr = Unix.inet6_addr_any in
-          (Xapi_http.bind (Unix.ADDR_INET (addr, Constants.http_port)), ":::443")
-        with _ ->
-          (* No. *)
-          let addr = Unix.inet_addr_any in
-          (Xapi_http.bind (Unix.ADDR_INET (addr, Constants.http_port)), "443")
-      )
-    | Some ip -> (
-        info "Starting new server (listening on %s)" ip ;
-        let addr = Unix.inet_addr_of_string ip in
-        let sockaddr = Unix.ADDR_INET (addr, Constants.http_port) in
-        ( Xapi_http.bind sockaddr
-        , match Unix.domain_of_sockaddr sockaddr with
-          | Unix.PF_INET6 ->
-              "::1:443"
-          | _ ->
-              "127.0.0.1:443" )
-      )
-  in
-  Http_svr.start Xapi_http.server socket ;
-  management_interface_server := socket :: !management_interface_server ;
-  restart_stunnel ~__context ~accept ;
-  if Pool_role.is_master () && !listening_all then
-    (* NB if we synchronously bring up the management interface on a master with a blank
-       		   database this can fail... this is ok because the database will be synchronised later *)
-    Server_helpers.exec_with_new_task "refreshing consoles" (fun __context ->
-        Dbsync_master.set_master_ip ~__context ;
-        Dbsync_master.refresh_console_urls ~__context)
+  let current_mode = ref Off
+
+  let update ~__context next_mode =
+    update' ~__context (!current_mode, next_mode) ;
+    current_mode := next_mode
+end
+
+(* High-level interface *)
+
+let reconfigure_stunnel = Stunnel.reconfigure_stunnel
 
 let change interface primary_address_type =
   Xapi_inventory.update Xapi_inventory._management_interface interface ;
   Xapi_inventory.update Xapi_inventory._management_address_type
     (Record_util.primary_address_type_to_string primary_address_type) ;
   update_mh_info interface
 
+let himn = ref None
+
+let management_m = Mutex.create ()
+
+let next_server_mode ~mgmt_enabled =
+  let localhost = "127.0.0.1" in
```

Comment:
Is this correct when we are using IPv6?

Reply:
This is an interesting question. We have experimental and incomplete management support for IPv6, and I'm hesitant to change things like this now. I think it should be reviewed if/when the time comes to put effort into making IPv6 support first class.